File: 2025-09-September-GFG-POTD/September 13 - Minimum Cost to cut a board into squares/Solution.java
class Solution {
    public static int minCost(int n, int m, int[] x, int[] y) {
        // Sort in descending order
        Arrays.sort(x);
        Arrays.sort(y);
        reverse(x);
        reverse(y);

        int i = 0, j = 0; // pointers for x and y
        int verticalSegments = 1, horizontalSegments = 1;
        int cost = 0;

        while (i < x.length && j < y.length) {
            if (x[i] >= y[j]) {
                // take vertical cut
                cost += x[i] * horizontalSegments;
                verticalSegments++;
                i++;
            } else {
                // take horizontal cut
                cost += y[j] * verticalSegments;
                horizontalSegments++;
                j++;
            }
        }

        // Remaining vertical cuts
        while (i < x.length) {
            cost += x[i] * horizontalSegments;
            i++;
        }

        // Remaining horizontal cuts
        while (j < y.length) {
            cost += y[j] * verticalSegments;
            j++;
        }

        return cost;
    }

    private static void reverse(int[] arr) {
        for (int i = 0, j = arr.length - 1; i < j; i++, j--) {
            int temp = arr[i];
            arr[i] = arr[j];
            arr[j] = temp;
        }
    }
}

File: 2025-09-September-GFG-POTD/September 13 - Minimum Cost to cut a board into squares/Solution.py
class Solution:
    @staticmethod
    def minCost(n, m, x, y):
        # Sort in descending order
        x.sort(reverse=True)
        y.sort(reverse=True)

        i = j = 0
        vertical_segments = 1
        horizontal_segments = 1
        cost = 0

        while i < len(x) and j < len(y):
            if x[i] >= y[j]:
                # vertical cut
                cost += x[i] * horizontal_segments
                vertical_segments += 1
                i += 1
            else:
                # horizontal cut
                cost += y[j] * vertical_segments
                horizontal_segments += 1
                j += 1

        while i < len(x):
            cost += x[i] * horizontal_segments
            i += 1

        while j < len(y):
            cost += y[j] * vertical_segments
            j += 1

        return cost
